sql/index.ts

```diff
@@ -132,12 +132,123 @@ export class Sql {
   }
 
   public asString(castableToString: string) {
-    if (this.dialect === "postgres") {
+    if (this.dialect === "postgres" || this.dialect === "redshift") {
       return `cast(${castableToString} as varchar)`;
     }
     return `cast(${castableToString} as string)`;
   }
 
+  // Surrogate keys.
+
+  public surrogateKey(columnNames: string[]) {
+    const columnsAsStrings = columnNames.map(id => this.asString(id)).join(`,`);
+    if (this.dialect === "standard") {
+      return this.asString(`farm_fingerprint(concat(${columnsAsStrings}))`);
+    }
+    if (
+      this.dialect === "postgres" ||
+      this.dialect === "redshift" ||
+      this.dialect === "snowflake"
+    ) {
+      return this.asString(`md5(concat(${columnsAsStrings}))`);
+    }
+    if (this.dialect === "mssql") {
+      return this.asString(`hashbytes("md5", (concat(${columnsAsStrings})))`);
+    }
+  }
+
+  // window function
+
+  public windowFunction(
+    name: string,
+    value: string,
+    ignoreNulls: boolean = false,
+    windowSpecification?: {
+      partitionFields?: string[];
+      orderFields?: string[];
+      frameClause?: string;
+    }
+  ) {
+    const partitionFieldsAsString = windowSpecification.partitionFields
+      ? [...windowSpecification.partitionFields].join(`, `)
+      : "";
+    const orderFieldsAsString = windowSpecification.orderFields
+      ? [...windowSpecification.orderFields].join(`, `)
+      : "";
+
+    if (this.dialect === "standard" || this.dialect === "mssql" || this.dialect === "snowflake") {
+      return `${name}(${value} ${ignoreNulls ? `ignore nulls` : ``}) over (${
+        windowSpecification.partitionFields ? `partition by ${partitionFieldsAsString}` : ``
+      } ${windowSpecification.orderFields ? `order by ${orderFieldsAsString}` : ``} ${
+        windowSpecification.frameClause ? windowSpecification.frameClause : ``
+      })`;
+    }
+
+    if (this.dialect === "redshift") {
+      // For some window functions in Redshift, a frame clause is required
+      const requiresFrame = [
```

Comment:
AFAICT these lists are identical here and for postgres, so pull it out into a constant before this `if` and you can use it in both places